packages/mui-base/src/utils/useLatest.ts

```diff
@@ -14,12 +14,7 @@ import * as React from 'react';
  * - [useLatest API](https://mui.com/base-ui/api/use-latest/)
  */
 export default function useLatest<T>(value: T, deps?: React.DependencyList) {
-  const ref = React.useRef<T>(value);
-
-  React.useEffect(() => {
-    ref.current = value;
-    // eslint-disable-next-line react-hooks/exhaustive-deps
-  }, deps ?? [value]);
-
-  return ref;
+  // eslint-disable-next-line react-hooks/exhaustive-deps
+  const current = React.useMemo(() => ({ current: value }), deps ?? [value]);
```

Comment:
The `useLatest` uses refs specifically to avoid rerendering all the list items whenever the selected/highlighted option changes. It results in a massive performance increase.
If `useLatest` uses `useMemo`, it essentially becomes unnecessary. This should be the last resort if nothing else could be done to resolve the bug while keeping the performant implementation.

The problem here lies in the `listItem` seeing stale `selected` value. I'd appreciate if you're able to investigate and debug it more deeply.